ext/standard/string.c

```diff
@@ -6331,6 +6331,38 @@ PHP_FUNCTION(utf8_decode)
 }
 /* }}} */
 
+/* {{{ proto bool str_truncate(string& str, int new_len)
+   Truncates the string to the new length. */
+PHP_FUNCTION(str_truncate)
+{
+	/*
+	 * Note that the proto specifies the str as a string reference, however,
+	 * internally speaking, it isn't. It's specified as a reference just to
+	 * reflect the fact the string would be changed upon calling this
+	 * function.
+	 *
+	 * This function is provided as a far more sufficient alternative of
+	 * truncating a large string to a still-very-large string.
+	 *
+	 * For example,
+	 * str_truncate($str, 1024 * 1024) vs $str = substr($str, 0, 1024 * 1024)
+	 * Obviously the former works far more sufficiently than the latter.
+	 */
+
+	zend_string *str;
+	zend_long new_len;
+
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_STR(str)
+		Z_PARAM_LONG(new_len)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (new_len < ZSTR_LEN(str) && new_len >= 0) {
+		ZSTR_LEN(str) = new_len;
```

Comment:
This just sets the length, but will not reduce the allocation of the string. IIRC we have a zend_string_truncate API.

Reply:
Also, throw a warning on invalid value (e.g. less than 0, greater than current len)

Reply:
@nikic We have zend_string_truncate(), but it'll always [copy the string](https://github.com/php/php-src/blob/php-7.3.0beta3/Zend/zend_string.h#L234), which would defeat the purpose of str_truncate().